### DISCUSSION TITLE
Future plans

### DISCUSSION THREAD

**Discussion by pyscripter** [TurboPack/SynEdit#126] [General]
The next two areas I am planning to work on are #26 and #27.

a) Unicode fixes (#26)

I plan to use DirectWrite for rendering text but also measurement and cursor positioning.   See [here](https://en.delphipraxis.net/topic/6054-choice-of-windows-text-rendering-engines/?tab=comments#comment-52627) for an explanation of the reasons.  

It will be a major rewrite.  The painting code, PaintTextLines, is about 1000 lines long.   I would expect that code to become much simpler.   Gone will be the weird stuff such as ExpandAtWideGlyphs.

The downside is that Direct2D and DirectWrite are not available on Windows XP.   I could keep the GDI stuff and switch to DirectWrite only when it is available.  But I do not want to do that.  It will defeat the purpose of simplifying the code base and it would be hard to maintain.  @MShark67 @JaFi-cz. Do you think XP compatibility is important?

b) Multi-cursor, multi-selection support (#27)

Some initial work has been done and it is available in the "multicursor" branch.  But it still needs quite a bit of work.

**Comment by JaFi-cz** [comment]
Form me, Windows XP support is still important.
When I publish test with new Delphi, where lowest Windows limit was set above Windows XP, I got lot of complains.
I understand, it's additional work for you to maintain both version of code, but Windows XP are still alive.

**Reply by pyscripter** [reply]
[About 1% of desktop/laptop users use Windows XP](https://www.netmarketshare.com/operating-system-market-share.aspx?options=%7B%22filter%22%3A%7B%22%24and%22%3A%5B%7B%22deviceType%22%3A%7B%22%24in%22%3A%5B%22Desktop%2Flaptop%22%5D%7D%7D%5D%7D%2C%22dateLabel%22%3A%22Trend%22%2C%22attributes%22%3A%22share%22%2C%22group%22%3A%22platformVersion%22%2C%22sort%22%3A%7B%22share%22%3A-1%7D%2C%22id%22%3A%22platformsDesktopVersions%22%2C%22dateInterval%22%3A%22Monthly%22%2C%22dateStart%22%3A%222019-11%22%2C%22dateEnd%22%3A%222020-10%22%2C%22segments%22%3A%22-1000%22%7D).

These users have not updated their OS for 20 years, so they must be quite happy with what they got.   Why would they care about updating your product?   Current versions of your software support Windows XP, future versions will not.  Your Win XP users/customers can carry on using the current version of your product.

**Reply by pyscripter** [reply]
[VS code for instance does not support Win XP](https://docs.microsoft.com/en-us/answers/questions/314855/visual-studio-code-for-windows-xp.html).

**Reply by JaFi-cz** [reply]
So why are you asking if you are decided?  ;-)

**Reply by pyscripter** [reply]
Good question.

What I have decided is that if I were to work on #26 I would not keep both GDI and DirectWrite in the code base.  The changes I have in mind are almost like writing Synedit from scratch and I just could not handle the complexity of keeping GDI and adding the DirectWrite.   Having both in, would also defeat part of the motivation, which is to simplify the code base.   In the early days of SynEdit there were tens, even hundreds of contributors.  Maybe that was part of the problem, but the code became almost unmaintainable.

What I have not decided is whether it is worth doing it.  You and Mark are the only developers that contribute to this project with features, support, bug fixing, testing etc.  Your help would be crucial.  If you were to tell me that "if you drop XP I am out and I will stick to the current version", I would just not bother.

**Reply by JaFi-cz** [reply]
You asked about WinXP support. I told you my opinion. If you leave support of WinXP, it doesn't change my mind to contribute and use current SynEdit version. I will publish last version with Windows XP support and next version will be without it. 
But if I would choose of new GDI handling or multicursot, I would vote for multicursor. There will be lot of work to tune it up.

**Reply by pyscripter** [reply]
I agree that multi-cursor multi-selection is more useful.   But it makes sense to deal with the lower level stuff first.  For example the Track changes bar was much easier to implement after the redesigned gutter and the new undo system.

**Comment by pyscripter** [comment]
By the way Marry Christmas to everyone.

**Comment by pyscripter** [comment]
One more question regarding multi-caret and multi-selection (please see https://wiki.freepascal.org/ATSynEdit#Multi-carets to get the idea.  VS-code works similarly).

If similar functionality is implemented in SynEdit, does one still need the different selection modes (column and line)?

**Reply by JaFi-cz** [reply]
I never used Line mode. I don't see usage for it, cause it can be easily done with standard selection. But column mode is still useful. Maybe pasting become obsolete, but column mode will always used for copy or delete. I personally use it often when I work with sets of data in text files. 
I see another usage of column mode - e.g. UpperCase/LowerCase column of data. Multiedit mode isn't able to replace it.
In my editor I have sort based on selected column implemented. But this can be handled by the other way.

**Reply by pyscripter** [reply]
> Multiedit mode isn't able to replace it

Why is that?  Have you read in the link how copy and past works?
For example in VS Code

<img width="102" alt="image" src="https://user-images.githubusercontent.com/1311616/147389541-6254bf83-c6e2-4e91-93a1-177f2ffda3e9.png">
Add cursors at the beginning of the lines and select lower
<img width="101" alt="image" src="https://user-images.githubusercontent.com/1311616/147389561-2dd4b551-dec0-481d-97ab-fc43c6602080.png">
type upper
<img width="102" alt="image" src="https://user-images.githubusercontent.com/1311616/147389584-6e8865aa-ae13-4471-b293-60194c8aaebf.png">
Select Upper
<img width="102" alt="image" src="https://user-images.githubusercontent.com/1311616/147389601-76ce5a28-e51a-4e75-9ab5-da11aa45bee3.png">
Apply Upper case command.
<img width="98" alt="image" src="https://user-images.githubusercontent.com/1311616/147389624-3edd2ffa-242c-47ea-bbed-7268da397dc7.png">

Note that VS code does not support column mode.

**Reply by JaFi-cz** [reply]
OK, Won't be there any performance problems to work with thousands lines at once?

**Reply by pyscripter** [reply]
Well you have a point.  VS Code has a limit of 10000 cursors.   But it gets very slow when say you type with more than 3-4000 cursors.  Selecting text with 10000 cursors is quite fast though.
Similarly with Cuda text.
I wonder whether the feature can be implemented more efficiently in SynEdit.  Will see.

**Comment by MShark67** [comment]
XP Support is only slightly important to me as I haven't had that issue pop up lately.  Of course, it's often that when something is removed that you find out someone really needs it lol.  I definitely approve of code simplification.  I've only used multi-cursor editing a few times, but it's definitely a cool feature.  One thing I'd like to see is a multi-line selection option that didn't highlight all the way to the right edge of the screen, but really it's a very minor point.  The recent improvements have all been stellar and much appreciated!

**Reply by pyscripter** [reply]
> multi-line selection option that didn't highlight all the way to the right edge of the screen

Delphi and SynEdit are the exception here rather than the norm.  It would be easy to switch, but I just don't like introducing new options.

**Reply by MShark67** [reply]
It's definitely a "wish list" item for me, but I can understand keeping with a standard (although it seems like some of my usability suggestions have been met with "that's how vscode does it", so maybe changing to be more standard isn't such a bad thing...)

**Reply by JaFi-cz** [reply]
I met this request once only, but for my opinion it isn't important if line is highlighted to the end of window or end of line. 
Highlight to end of line can be confusing for empty lines, cause you won't see if line is selected there

**Reply by MShark67** [reply]
JaFi, that's a good point. I checked out VSCode and it appears they do the end of line selection as if there was a hidden char there. I wonder if it's for that reason.  I've had someone complain about the full line selects as it hides what whitespace is there, so there are pros and cons both ways I think.

**Reply by pyscripter** [reply]
What should we do with this?  Switch, make it optional or keep it as is?

**Reply by MShark67** [reply]
I'd vote for either an option or to just make it the new standard (but I totally understand keeping it "as is" as well.)  I do like the way VSCode does it, I noticed they've moved the behavior to outlook as well.

**Comment by pyscripter** [comment]
Finished and committed the painting code.
Sample output:
![image](https://user-images.githubusercontent.com/1311616/150265474-b03c934b-8732-4895-ae8e-b013454a7df1.png)
Chinese properly spaced, surrogate pairs and color emojis
![image](https://user-images.githubusercontent.com/1311616/150265824-cbf6652a-bbbc-457c-9498-5f09f9b60423.png)
Bidirectional text properly selected.
![image](https://user-images.githubusercontent.com/1311616/150266107-fdccabd2-c766-4ec3-a674-13abdf91e63b.png)

**Reply by MShark67** [reply]
Looks amazing!  The quality of the characters is perfect (or at least I can't notice any differences between this and the non-dwrite version.)  If I load a big html or pas file without a highlighter selected, I see the indent level drawing and vertical scroll performance seems to be affected adversely by whatever it's doing.  If I load a text file without indents it seems to scroll very quickly.

**Reply by MShark67** [reply]
Tracked it down to 
    if CodeFolding.IndentGuides and not WordWrap then
      DrawIndentGuildes;
I think it just needs a "if UseCodeFolding" check.

**Reply by pyscripter** [reply]
Is the problem that is slow or that it is drawing them?  I though I would decouple Indentation Guides form Code Folding.  There is no reason of not having them without Code Folding.

DrawIdentGuides is the slowest part of the drawing.  It can take up 80% of the painting time.  The problem is that DirectWrite is very slow in drawing non solid lines.  If the lines are solid it is much faster.

There are a couple of options:
-  Draw Indentation lines with GDI like we did before
- Replace them with solid lines  It looks good if it is thinner and with TColors.Lightgray

Could you please try the speed of your htmll file with solid lines?  
In TSynDWrite.DottedStrokeStyle replace D2D1_DASH_STYLE_DOT with D2D1_DASH_STYLE_SOLID>

**Reply by MShark67** [reply]
I was testing in my fairly simple personal editor which doesn't use code folding or syntax highlighting, so I was surprised to see the indent guides drawing by default (but I can easily change that option of course.)  The issue was stuttering while "thumbtrack" scrolling vertically in a large file with lots of indent levels.  Changing to SOLID instead of DOT does indeed make a big difference!  I actually would prefer thin solid lines over dotted ones.

**Reply by pyscripter** [reply]
> does indeed make a big difference!

Does this mean that now there are no noticeable delays?

**Reply by pyscripter** [reply]
![image](https://user-images.githubusercontent.com/1311616/150386843-930d9f61-821e-455e-b9ac-3cc318a13761.png)
How does this look?

**Reply by JaFi-cz** [reply]
Font looks great, but I would prefer thinner guide lines.

**Reply by MShark67** [reply]
> Does this mean that now there are no noticeable delays?
The difference between having the guide painting on and off is noticeable; but I don't think it's a big issue.  If I turn the feature off scrolling is pretty smooth.  If I turn it on, I get some minor scrolling speed stuttering.

**Reply by pyscripter** [reply]
@JaFi-cz The lines above are one pixel thick.  I don't think you can get them thinner.

I have found a way to draw the dotted lines much faster (draw to a bitmap and then render the bitmap multiple times).   We can always switch to solid lines, or make it optional.  This is what it looks like with LightGray color.

![image](https://user-images.githubusercontent.com/1311616/150445497-1985a323-e4d5-4ea5-8eb5-9647393563e2.png)


@MShark67 Could you please check the performance again with your html file.  Here performance is good, but my machine is very fast.   

Please note that quite a few things may not work correctly.  There is still a long way to go, but the hardest part is done.

**Reply by pyscripter** [reply]
Another option is turn off anti-aliasing while drawing indentation lines.  This is what it looks like with color clGray.
![image](https://user-images.githubusercontent.com/1311616/150529337-5fd162e3-277b-4f2a-a57b-65576def58d5.png)
Quite close to what we have today.  And should be much faster.
And Aliased straight lines with LightGray color
![image](https://user-images.githubusercontent.com/1311616/150530175-6300309a-d38a-4859-b113-b0c35dec634d.png)

**Reply by MShark67** [reply]
The performance of the new bitmap version is very good, no longer an issue at all.  I notice that the lines overlap slightly causing a darker dot between lines.  I don't see it in your screen shots so maybe you've already fixed it or it's more pronounced on large font sizes.  I can (kind of) fix it by reducing the bitmap height (or the draw rectangle) by one pixel.  Here's a screenshot that shows it with a huge font just to make it clear:
![SynEdit1](https://user-images.githubusercontent.com/38496584/150539025-4cefadf6-e8d9-4837-992d-8663dc387d11.png)

**Reply by JaFi-cz** [reply]
Version of gudes without antialiassing looks much better. Thank you.
Can you make font antialliasing optional please? Some of my users require it.

**Reply by pyscripter** [reply]
> Can you make font antialliasing optional please? Some of my users require it.

Now there is the FontQuality property, which also affects the DWrite stuff.  Let me finish everything and then we can think about what rendering options are needed.

Dotted or straight lines?  I think I like better the aliased straight lines with the appropriate choice of color.  Maybe this could also be an option.

**Reply by JaFi-cz** [reply]
> Dotted or straight lines? I think I like better the aliased straight lines with the appropriate choice of color. Maybe this could also be an option.

Straight lines without aliasing looks good. I would prefer speed than dotted lines.

**Comment by pyscripter** [comment]
Discussion moved to a new topic #141
